packages/core/client.js

```diff
@@ -1,56 +1,59 @@
 const config = require('./config')
-const BugsnagReport = require('./report')
-const BugsnagBreadcrumb = require('./breadcrumb')
-const BugsnagSession = require('./session')
-const { map, includes, isArray } = require('./lib/es-utils')
-const inferReleaseStage = require('./lib/infer-release-stage')
+const Event = require('./event')
+const Session = require('./session')
+const Breadcrumb = require('./breadcrumb')
+const { filter, map, includes, isArray, keys } = require('./lib/es-utils')
 const isError = require('./lib/iserror')
 const some = require('./lib/async-some')
-const runBeforeSend = require('./lib/run-before-send')
+const createCallbackRunner = require('./lib/async-callback-runner')
+const noop = () => {}
+const metadataDelegate = require('./lib/metadata-delegate')
 
-const LOG_USAGE_ERR_PREFIX = `Usage error.`
-const REPORT_USAGE_ERR_PREFIX = `Bugsnag usage error.`
+const LOG_USAGE_ERR_PREFIX = 'Usage error.'
+const REPORT_USAGE_ERR_PREFIX = 'Bugsnag usage error.'
 
-class BugsnagClient {
-  constructor (notifier) {
-    if (!notifier || !notifier.name || !notifier.version || !notifier.url) {
-      throw new Error('`notifier` argument is required')
-    }
-
-    // notifier id
-    this.notifier = notifier
+class Client {
+  constructor (configuration, schema = config.schema, notifier) {
+    this._notifier = notifier
 
-    // configure() should be called before notify()
-    this._configured = false
+    // if the facade interface is sitting in front of this class,
+    // it should update this value so that generated stacktraces
+    // have the correct amount of frames removed
+    this._depth = 0
 
     // intialise opts and config
-    this._opts = {}
-    this.config = {}
+    this._opts = configuration
+    this._config = {}
+    this._schema = schema
 
-    // // i/o
-    this._delivery = { sendSession: () => {}, sendReport: () => {} }
-    this._logger = { debug: () => {}, info: () => {}, warn: () => {}, error: () => {} }
-
-    // plugins
-    this._plugins = {}
+    this.__delivery = { sendSession: noop, sendEvent: noop }
+    this.__logger = { debug: noop, info: noop, warn: noop, error: noop }
+    this.__sessionDelegate = { startSession: () => this, pauseSession: noop, resumeSession: noop }
 
     this._session = null
+    this._pausedSession = null
+
+    this._breadcrumbs = []
+    this._context = undefined
+    this._metadata = {}
+    this._user = {}
+
+    this._cbs = {
```

Comment:
It's a data structure where the callbacks are stored.

It has a terse name because property names do not get shortened when you minify JS. Perhaps I should add a comment?